bugbug/model_calibration.py

```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import numpy as np
+from sklearn.isotonic import IsotonicRegression
+from sklearn.metrics import mean_squared_error
+from sklearn.model_selection import train_test_split
+
+
+class IsotonicRegressionCalibrator:
+    def __init__(self, model):
+        self.model = model
+        self.calibrator = IsotonicRegression()
+
+    def train_test_split(self, X, y, test_size=0.2, random_state=42):
+        X_train, X_test, y_train, y_test = train_test_split(
+            X, y, test_size=test_size, random_state=random_state
+        )
+        return X_train, X_test, y_train, y_test
+
+    def fit(self, X_train, y_train):
+        X_train, X_val, y_train, y_val = self.train_test_split(X_train, y_train)
+        self.model.fit(X_train, y_train)
+        mse_before = mean_squared_error(y_val, self.model.predict(X_val))
+        print(f"MSE of model before calibration : {mse_before:.4f}")
+
+        self.calibrate(X_val, y_val)
+        mse_after = mean_squared_error(y_val, self.predict(X_val))
+        print(f"MSE of model after calibration : {mse_after:.4f}")
+
+    def predict(self, X):
+        p_pred = self.calibrator.predict(self.model.predict_proba(X)[:, 1])
+        # transforming predictions back to class labels
+        p_pred = p_pred.flatten()
+        y_pred = np.where(p_pred > 0.5, 1, 0)
+
+        return y_pred
+
+    def predict_proba(self, X_val):
+        return self.calibrator.transform(self.model.predict_proba(X_val)[:, 1])
```

Comment:
Are we returning the expected shape here?

Reply:
> > @suhaibmujahid so i used sklearn.calibration.CalibratedClassifierCV instead of IsotonicRegression model and for now set  self.calculate_importance = False  (i figured out that IsotonicRegressionCalibrator model type is not currently supported by the TreeExplainer function in the shap library thats why i was getting the "Model type not yet supported by TreeExplainer: <class 'bugbug.model_calibration.IsotonicRegressionCalibrator'>" error ). The training is now done without any errors!
> 
> Did you check the following part of the Model class? Isn't this duplicating the cross-validation work?
> 
> https://github.com/mozilla/bugbug/blob/2075ddb4eafeddc558974bf910a7264a15d74f80/bugbug/model.py#L367-L382

thats something i didnt even touch so I didnt notice but let me check it out!

Reply:
But this applies to all models with the `cross_validation_enabled` flag enabled.